craft_parts/utils/partition_utils.py

```diff
@@ -118,58 +115,123 @@ def _validate_partition_naming_convention(partitions: Sequence[str]) -> None:
+    lines: list[str] = []
+    for conflictings in filtered_conflictings_partitions.values():
+        conflictings_list = list(conflictings)
+        conflictings_list.sort()
+        lines.append(f"- {str(conflictings_list)[1:-1]}")
+
+    msg = "\n".join(lines)
+
+    raise errors.FeatureError(
+        message=f"Following partitions conflicts:\n{msg}",
+        details="Hyphens and slashes are converted to underscores to associate partitions names with environment variables. 'foo-bar' and 'foo/bar' would result in environment variable FOO_BAR.",
+    )
+
+
+def detect_conflicts(partitions: Sequence[str]) -> dict[str, set[str]]:
+    """Detect and return every conflicts between partitions.
 
-    For example, `foo` conflicts in ['default', 'foo', 'foo/bar'].
-    Assumes partition names are valid.
+    Rules:
+      1: partition name must not conflict with namespace (or sub-namespace) names
+        - `foo` and 'foo/bar' conflicts
+        - `foo/bar` and 'foo/bar/baz' conflicts
+      2: environment variables derived from partition names must not conflicts
+        - `foo/bar-baz` and `foo/bar/baz` conflicts (would be converted to FOO_BAR_BAZ)
+        - `foo/bar/baz-qux` and `foo/bar-baz/qux` conflicts (would be converted to FOO_BAR_BAZ_QUX)
 
-    :raises FeatureError: for namespace conflicts
     """
-    namespaced_partitions: set[str] = set()
-    regular_partitions: set[str] = set()
+    conflictings_partitions: dict[str, set[str]] = defaultdict(set)
+
+    # Rule #1: Detect conflicts between partitions and namespaces partitions
+    for candidate_partition in partitions:
+        for partition in partitions:
+            if candidate_partition == partition:
+                continue
+            if namespace_conflicts(candidate_partition, partition):
+                if not conflictings_partitions[partition]:
+                    conflictings_partitions[partition].add(partition)
+                conflictings_partitions[partition].add(candidate_partition)
+
+    # Rule #2: Detect conflicts between environment variable names
+    underscored_partitions: dict[str, str] = {}
 
-    # sort partitions
     for partition in partitions:
-        if _is_valid_partition_name(partition):
-            regular_partitions.add(partition)
+        underscored = partition.translate({ord("-"): "_", ord("/"): "_"})
+
+        for known_partition, known_underscored in underscored_partitions.items():
+            if env_var_conflicts(underscored, known_underscored):
+                if not conflictings_partitions[known_partition]:
+                    conflictings_partitions[known_partition].add(known_partition)
+                conflictings_partitions[known_partition].add(partition)
+
+        underscored_partitions[partition] = underscored
+
+    return conflictings_partitions
+
+
+def split_and_detect(a: str, b: str, separator: str) -> bool:
+    """Split candidates with the separator and check if the shorter one is a subset of the other."""
```

Comment:
If this is performed on a single separator, can we just check if one string `startswith` the other?

Reply:
That was the initial implementation but I failed to make it work because:

- `foo` and `foo/bar` must conflict (the previous behavior)
- `foo/bar-baz` and `foo/bar/baz` must now also conflict
- `foo/bar` and `foo/bar/baz` must now also conflict
- `foo/bar` and `foo/barbaz`  must **NOT** conflict (here using `startswith` would detect a conflict)
 
 I just simplified and fixed it though. Let me know what you think.